### ISSUE TITLE
Unnecessary extra memory allocation when using circular padding

### ISSUE THREAD

**Issue by cth7** [pytorch/pytorch#39256]
## 🐛 Bug

When using circular padding a significant amount of memory wants to be allocated that doesn't appear to be used.

## To Reproduce

Following script should reproduce.

```python
import torch
import torch.nn as nn


def torch_memory(device):
    # Checks and prints GPU memory
    print(f'{torch.cuda.memory_allocated(device)/1024/1024:.2f} MB USED')
    print(f'{torch.cuda.memory_reserved(device)/1024/1024:.2f} MB RESERVED')
    print(f'{torch.cuda.max_memory_allocated(device)/1024/1024:.2f} MB USED MAX')
    print(f'{torch.cuda.max_memory_reserved(device)/1024/1024:.2f} MB RESERVED MAX')
    print('')


print(torch.__version__)

device = 0
device = torch.device(device)

# Create input
x = torch.randn((1, 2, 24, 512, 512), dtype=torch.float32, device=device)

# Create conv layers
pmode = 'circular'
# pmode = 'zeros'
width = 64
conv0 = torch.nn.Conv3d(2, width, 3, padding=1, padding_mode=pmode).to(device)
conv1 = torch.nn.Conv3d(width, width, 3, padding=1, padding_mode=pmode).to(device)
conv2 = torch.nn.Conv3d(width, 2, 3, padding=1, padding_mode=pmode).to(device)
torch_memory(device.index)

with torch.no_grad():
    x = conv0(x)
    print('After first conv layer.')
    torch_memory(device.index)

    x = conv1(x)
    print('After second conv layer.')
    torch_memory(device.index)  # A lot more memory is reserved here

    x = conv2(x)
    print('After third conv layer.')
    torch_memory(device.index)
```
## Expected behavior

When using circular padding I get this output. The extra unnecessary memory allocation is particularly significant after the second convolution layer. It uses a maximum of 4888 MB, which makes sense because circular padding concatenates slices into a new array. But it reserves 8510 MB, which doesn't make sense to me and gives me OOM errors during training with more conv layers.
```
1.6.0.dev20200528
48.45 MB USED
50.00 MB RESERVED
48.45 MB USED MAX
50.00 MB RESERVED MAX

After first conv layer.
1536.45 MB USED
1906.00 MB RESERVED
1745.05 MB USED MAX
1906.00 MB RESERVED MAX

After second conv layer.
1536.45 MB USED
6562.00 MB RESERVED
4888.45 MB USED MAX
8510.00 MB RESERVED MAX

After third conv layer.
48.45 MB USED
6562.00 MB RESERVED
4888.45 MB USED MAX
8510.00 MB RESERVED MAX
```
When padding with zeros, I get this output. Here it, both, uses and reserves an amount that makes sense.
```
1.6.0.dev20200528
48.45 MB USED
50.00 MB RESERVED
48.45 MB USED MAX
50.00 MB RESERVED MAX

After first conv layer.
1536.45 MB USED
1696.00 MB RESERVED
1692.64 MB USED MAX
1696.00 MB RESERVED MAX

After second conv layer.
1536.45 MB USED
3232.00 MB RESERVED
3180.64 MB USED MAX
3232.00 MB RESERVED MAX

After third conv layer.
48.45 MB USED
3232.00 MB RESERVED
3180.64 MB USED MAX
3232.00 MB RESERVED MAX
```

## Environment

 - PyTorch Version (e.g., 1.0): Nightly, because there were changes since 1.5 that would otherwise have made the memory outputs confusing. I also noticed this in versions 1.1 and 1.4, though the memory numbers were slightly different.
 - OS (e.g., Linux): Windows 10 and CentOS 7
 - How you installed PyTorch (`conda`, `pip`, source): conda
 - Build command you used (if compiling from source): N/A
 - Python version: 3.8.3
 - CUDA/cuDNN version: CUDA 10.1, cuDNN 7.6
 - GPU models and configuration: GTX 1080 TI and V100
 - Any other relevant information:

## Additional context

I also tried 2D convs but the extra memory allocations didn't seem as dramatic, despite being a dimension smaller...

cc @albanD @mruberry

**Comment by chcodes** [comment]
Currently, torch.nn.functional._pad_circular uses several torch.cat to create the padded array. It seems inefficient if this allocates new arrays for every torch.cat? Still doesn't explain why it wants to reserve so much more than it uses though.

I've written an alternative that first creates a new array at the padded array size, puts the original array into the padded array at the right location, and then uses manual slicing on the padded array to do the circular padding. I'm not 100% sure on the underlying memory mechanics or if this makes anything inefficient, but so far it appears to use/reserve the right amount. Also, have no idea what the implications would be on the backward pass or if it's something I need to worry about. I still need to check a few things but would be happy to hear anyone's thoughts on this. It looks something like this:
```python
def new_circ_pad(input, padding):    
    shape = input.shape
    assert len(shape[2:])*2 == len(padding)
    ndim = len(shape[2:])
    assert ndim in [1, 2, 3]

    # Get shape of padded array
    new_shape = shape[:2]
    for a, size in enumerate(shape[2:]):
        new_shape += (size + padding[a*2] + padding[a*2+1],)

    out = torch.empty(new_shape, dtype=input.dtype, layout=input.layout,
                      device=input.device)

    # Put original array in padded array
    if ndim == 1:
        out[..., padding[0]:-padding[1]] = input
    elif ndim == 2:
        out[..., padding[0]:-padding[1], padding[2]:-padding[3]] = input
    elif ndim == 3:
        out[..., padding[0]:-padding[1], padding[2]:-padding[3], padding[4]:-padding[5]] = input

    # Pad right side, then left side. Corners will be written more than once
    # when ndim > 1

    # Pad first dim
    out[:, :, :padding[0]] = out[:, :, -(padding[0] + padding[1]):-padding[1]]
    out[:, :, -padding[1]:] = out[:, :, padding[0]:(padding[0] + padding[1])]

    if len(padding) > 2:
        # Pad second dim
        out[:, :, :, :padding[2]] = out[:, :, :, -(padding[2] + padding[3]):-padding[3]]
        out[:, :, :, -padding[3]:] = out[:, :, :, padding[2]:(padding[2] + padding[3])]

    if len(padding) > 4:
        # Pad third dim
        out[:, :, :, :, :padding[4]] = out[:, :, :, :, -(padding[4] + padding[5]):-padding[5]]
        out[:, :, :, :, -padding[5]:] = out[:, :, :, :, padding[4]:(padding[4] + padding[5])]

    return out
```

**Comment by cth7** [comment]
Resolved by https://github.com/pytorch/pytorch/pull/39273.
